<a href="https://colab.research.google.com/github/thamarai-s15/ML_diabeties_model/blob/main/ML_model.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Task
Generate a machine learning model using Support Vector Classification with hyperparameter tuning using GridSearchCV and RandomizedSearchCV on the dataset provided in "/content/diabetes.csv".

## Load the data

### Subtask:
Load the dataset from the specified file into a pandas DataFrame.


**Reasoning**:
Import the pandas library and load the dataset from the specified file into a pandas DataFrame. Display the first 5 rows and the information summary.



In [ ]:
import pandas as pd

df = pd.read_csv('/content/diabetes.csv')
display(df.head())
display(df.info())

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


None

## Preprocess the data

### Subtask:
Handle any missing values and scale the features as necessary.


**Reasoning**:
Check for missing values, replace zero values in specific columns with NaN, impute missing values with the mean, separate features and target, and scale the features.



In [ ]:
print("Missing values before handling:")
print(df.isnull().sum())

cols_with_zeros = ['Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI']
for col in cols_with_zeros:
    df[col] = df[col].replace(0, np.nan)

print("\nMissing values after replacing zeros:")
print(df.isnull().sum())

from sklearn.impute import SimpleImputer
import numpy as np

imputer = SimpleImputer(missing_values=np.nan, strategy='mean')
df[cols_with_zeros] = imputer.fit_transform(df[cols_with_zeros])

print("\nMissing values after imputation:")
print(df.isnull().sum())

X = df.drop('Outcome', axis=1)
y = df['Outcome']

from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

display(X_scaled[:5])
display(y[:5])

Missing values before handling:
Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64

Missing values after replacing zeros:
Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64

Missing values after imputation:
Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64


array([[ 6.39947260e-01,  8.65108070e-01, -3.35182392e-02,
         6.65502121e-01, -3.34507888e-16,  1.66291742e-01,
         4.68491977e-01,  1.42599540e+00],
       [-8.44885053e-01, -1.20616153e+00, -5.29859033e-01,
        -1.74633816e-02, -3.34507888e-16, -8.52531177e-01,
        -3.65060778e-01, -1.90671905e-01],
       [ 1.23388019e+00,  2.01581340e+00, -6.95305964e-01,
         8.08793628e-16, -3.34507888e-16, -1.33283341e+00,
         6.04397318e-01, -1.05584152e-01],
       [-8.44885053e-01, -1.07465235e+00, -5.29859033e-01,
        -7.00428884e-01, -7.24388738e-01, -6.34211980e-01,
        -9.20762614e-01, -1.04154944e+00],
       [-1.14185152e+00,  5.03457822e-01, -2.68066914e+00,
         6.65502121e-01,  1.46550563e-01,  1.54897999e+00,
         5.48490910e+00, -2.04963989e-02]])

0    1
1    0
2    1
3    0
4    1
Name: Outcome, dtype: int64

**Reasoning**:
The previous code failed because numpy was not imported. I need to import numpy before using it.



In [ ]:
print("Missing values before handling:")
print(df.isnull().sum())

cols_with_zeros = ['Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI']
import numpy as np
for col in cols_with_zeros:
    df[col] = df[col].replace(0, np.nan)

print("\nMissing values after replacing zeros:")
print(df.isnull().sum())

from sklearn.impute import SimpleImputer

imputer = SimpleImputer(missing_values=np.nan, strategy='mean')
df[cols_with_zeros] = imputer.fit_transform(df[cols_with_zeros])

print("\nMissing values after imputation:")
print(df.isnull().sum())

X = df.drop('Outcome', axis=1)
y = df['Outcome']

from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

display(X_scaled[:5])
display(y[:5])

Missing values before handling:
Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64

Missing values after replacing zeros:
Pregnancies                   0
Glucose                       5
BloodPressure                35
SkinThickness               227
Insulin                     374
BMI                          11
DiabetesPedigreeFunction      0
Age                           0
Outcome                       0
dtype: int64

Missing values after imputation:
Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64


array([[ 6.39947260e-01,  8.65108070e-01, -3.35182392e-02,
         6.65502121e-01, -3.34507888e-16,  1.66291742e-01,
         4.68491977e-01,  1.42599540e+00],
       [-8.44885053e-01, -1.20616153e+00, -5.29859033e-01,
        -1.74633816e-02, -3.34507888e-16, -8.52531177e-01,
        -3.65060778e-01, -1.90671905e-01],
       [ 1.23388019e+00,  2.01581340e+00, -6.95305964e-01,
         8.08793628e-16, -3.34507888e-16, -1.33283341e+00,
         6.04397318e-01, -1.05584152e-01],
       [-8.44885053e-01, -1.07465235e+00, -5.29859033e-01,
        -7.00428884e-01, -7.24388738e-01, -6.34211980e-01,
        -9.20762614e-01, -1.04154944e+00],
       [-1.14185152e+00,  5.03457822e-01, -2.68066914e+00,
         6.65502121e-01,  1.46550563e-01,  1.54897999e+00,
         5.48490910e+00, -2.04963989e-02]])

0    1
1    0
2    1
3    0
4    1
Name: Outcome, dtype: int64

## Split the data

### Subtask:
Split the dataset into training and testing sets.


**Reasoning**:
Split the scaled feature data and the target variable into training and testing sets using train_test_split.



In [ ]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.2, random_state=42)

print("Shape of X_train:", X_train.shape)
print("Shape of X_test:", X_test.shape)
print("Shape of y_train:", y_train.shape)
print("Shape of y_test:", y_test.shape)

Shape of X_train: (614, 8)
Shape of X_test: (154, 8)
Shape of y_train: (614,)
Shape of y_test: (154,)


## Train and tune the model

### Subtask:
Train a Support Vector Classification model and tune its hyperparameters using GridSearchCV and RandomizedSearchCV.


**Reasoning**:
Train a Support Vector Classification model and tune its hyperparameters using GridSearchCV and RandomizedSearchCV.



In [ ]:
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV, RandomizedSearchCV

param_grid = {'C': [0.1, 1, 10, 100], 'gamma': [1, 0.1, 0.01, 0.001]}
grid_search = GridSearchCV(SVC(), param_grid, cv=5)
grid_search.fit(X_train, y_train)

param_dist = {'C': [0.1, 1, 10, 100], 'gamma': [1, 0.1, 0.01, 0.001]}
random_search = RandomizedSearchCV(SVC(), param_dist, n_iter=10, cv=5, random_state=42)
random_search.fit(X_train, y_train)

print("Best parameters for GridSearchCV:", grid_search.best_params_)
print("Best score for GridSearchCV:", grid_search.best_score_)

print("Best parameters for RandomizedSearchCV:", random_search.best_params_)
print("Best score for RandomizedSearchCV:", random_search.best_score_)

Best parameters for GridSearchCV: {'C': 100, 'gamma': 0.01}
Best score for GridSearchCV: 0.7768759163001466
Best parameters for RandomizedSearchCV: {'gamma': 0.01, 'C': 100}
Best score for RandomizedSearchCV: 0.7768759163001466


## Evaluate the model

### Subtask:
Evaluate the performance of the tuned model on the testing set.


**Reasoning**:
Evaluate the performance of the best model on the testing set by making predictions and calculating accuracy, precision, recall, and F1-score.



In [ ]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

best_model = grid_search.best_estimator_
y_pred = best_model.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)

print(f"Accuracy: {accuracy:.4f}")
print(f"Precision: {precision:.4f}")
print(f"Recall: {recall:.4f}")
print(f"F1-score: {f1:.4f}")

Accuracy: 0.7403
Precision: 0.6415
Recall: 0.6182
F1-score: 0.6296


## Summary:

### Data Analysis Key Findings

*   The dataset initially showed no missing values, but replacing zero values in 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', and 'BMI' with NaN revealed missing data. These missing values were successfully imputed using the mean of their respective columns.
*   Features were scaled using `StandardScaler` before splitting the data.
*   The dataset was split into 80% for training (614 samples) and 20% for testing (154 samples).
*   Both GridSearchCV and RandomizedSearchCV identified the same best hyperparameters for the SVC model: {'C': 100, 'gamma': 0.01}, achieving a cross-validation score of approximately 0.777 on the training data.
*   On the test set, the tuned SVC model achieved an Accuracy of 0.7403, Precision of 0.6415, Recall of 0.6182, and an F1-score of 0.6296.

### Insights or Next Steps

*   The performance on the test set is slightly lower than the cross-validation score, suggesting potential overfitting or that the test set is slightly different from the training set distribution. Further analysis could involve exploring different kernel functions for the SVC model or investigating the misclassified samples.
*   Consider exploring alternative classification algorithms or ensemble methods to potentially improve the model's performance, particularly the precision and recall scores.
